File: string/126.py
def check_balanced_expression(cad: str):
    stack = []

    for ch in cad:
        if len(stack) > 0 and stack[-1] == "[" and ch == "]":
            stack.pop()
            continue
        if len(stack) > 0 and stack[-1] == "(" and ch == ")":
            stack.pop()
            continue
        if len(stack) > 0 and stack[-1] == "{" and ch == "}":
            stack.pop()
            continue

        stack.append(ch)

    return len(stack) == 0


cad = "{[{}{}]}[()]"
print(check_balanced_expression(cad))
cad = "{()}[)"
print(check_balanced_expression(cad))

File: string/126.cpp
#include <bits/stdc++.h>

using namespace std;

bool checkBalancedExpression(const string& cad) {
    stack<char> st;

    for (const char& ch : cad) {
        if (!st.empty() && st.top() == '[' && ch == ']') {
            st.pop();
            continue;
        } else if (!st.empty() && st.top() == '(' && ch == ')') {
            st.pop();
            continue;
        } else if (!st.empty() && st.top() == '{' && ch == '}') {
            st.pop();
            continue;
        }

        st.push(ch);
    }

    return st.empty();
}

int main() {

    string cad = "{[{}{}]}[()]";
    cout << boolalpha << checkBalancedExpression(cad) << "\n";

    cad = "{()}[)";
    cout << boolalpha << checkBalancedExpression(cad);


    return 0;
}